whiteout/resolver.go

```diff
@@ -0,0 +1,92 @@
+package whiteout
+
+import (
+	"context"
+	"path/filepath"
+	"strings"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/indexer"
+	"github.com/quay/zlog"
+)
+
+var (
+	_ indexer.Resolver = (*Resolver)(nil)
+)
+
+type Resolver struct{}
+
+func (r *Resolver) Resolve(ctx context.Context, ir *claircore.IndexReport, layers []*claircore.Layer) *claircore.IndexReport {
+	// Here we need to check if any of the packages
+	// found are moot due to whiteouts.
+	ls := newLayerSorter(layers)
+	finalPackages := map[string]*claircore.Package{}
+	finalEnvironments := map[string][]*claircore.Environment{}
+	for pkgID, pkg := range ir.Packages {
+		packageDeleted := false
+		// First env.IntroducedIn?
+		packageLayer := ir.Environments[pkgID][0].IntroducedIn.String()
+		for fileLayer, f := range ir.Files {
+			// Check if it's a whiteout file, if it applies to the package's
+			// file and if the layer the whiteout file came from came after.
+			if f.Kind == claircore.FileKindWhiteout && ls.isChildOf(fileLayer, packageLayer) && fileIsDeleted(pkg.File, f.Path) {
+				packageDeleted = true
+				zlog.Debug(ctx).
+					Str("package name", pkg.Name).
+					Str("package file", pkg.File).
+					Str("whiteout file", f.Path).
+					Msg("package determined to be deleted")
+			}
+		}
+		switch packageDeleted {
+		case true:
+		default:
+			finalPackages[pkgID] = pkg
+			finalEnvironments[pkgID] = ir.Environments[pkgID]
+		}
+	}
+	ir.Packages = finalPackages
+	ir.Environments = finalEnvironments
+	return ir
+}
+
+// fileIsDeleted returns whether or not the filepath(fp) has been deleted
+// by the corresponding whiteoutPath. It follows the OCI spec for whiteouts:
+// https://github.com/opencontainers/image-spec/blob/main/layer.md#whiteouts
+func fileIsDeleted(fp, whiteoutPath string) bool {
+	var checkFile string
+	switch {
+	case filepath.Base(whiteoutPath) == ".wh..wh..opq":
+		// Special opaque case, "indicating that all siblings are hidden in the lower layer"
+		checkFile = filepath.Dir(whiteoutPath)
+	default:
+		checkFile = strings.ReplaceAll(whiteoutPath, ".wh.", "")
+	}
+
+	// There has to be a better way
```

Comment:
Perhaps take a look a the `filepath.IsRel` implementation and adapt it? This is probably not too far off

Reply:
I took a look at  this today and I think it's the least worst option considering:

1. We need path prefix not just string prefix
2. We're lucky enough that the paths are not relative (so the code can be simpler),
3. We could use filepath.Match() but the fact that we don't know how deep the match will be i.e `.wo.root` and `root/algo/something/interesting.txt`. This makes the pattern generation non-trivial / makes it less readable IMO